aioxmpp/ibb/service.py

```diff
@@ -0,0 +1,507 @@
+# License along with this program.  If not, see
+# <http://www.gnu.org/licenses/>.
+#
+########################################################################
+import asyncio
+import base64
+import contextlib
+import os
+
+import aioxmpp
+import aioxmpp.errors as errors
+import aioxmpp.callbacks
+import aioxmpp.service as service
+
+from . import xso as ibb_xso
+
+
+MAX_BLOCK_SIZE = (1<<16) - 1
+
+
+def _generate_session_id():
+    # XXX: should we use secrets.token
+    return base64.urlsafe_b64encode(os.urandom(12)).decode("us-ascii")
+
+
+class IBBTransport(asyncio.Transport):
+    """
+    The transport for IBB sessions.
+
+    .. note:: Never instantiate this class directly, all instances of
+              this class are created by :class:`~aioxmpp.ibb.IBBService`.
+
+    The following keys are supported for
+    :meth:`~asyncio.BaseTransport.get_extra_info`:
+
+       `block_size`
+           The maximal block size of data in a IBB stanza.
+
+       `peer_jid`
+          The JID of the peer.
+
+       `sid`
+          The session id of the unerlying IBB session.
+
+       `stanza_type`
+          The used stanza type.
+    """
+
+    def __init__(self, service, peer_jid, sid, stanza_type, block_size):
+        self._protocol = None
+        self._service = service
+        self._stanza_type = stanza_type
+        self._sid = sid
+        self._peer_jid = peer_jid
+        self._block_size = block_size
+        self._incoming_seq = 0
+        self._outgoing_seq = 0
+        self._closed = False
+        self._closing = False
+
+        self.set_write_buffer_limits()
+        self._write_buffer = b""
+        self._can_write = asyncio.Event()
+
+        self._reading_paused = False
+        self._input_buffer = []
+
+        self._write_task = asyncio.ensure_future(self._write_task_main())
+
+    def set_write_buffer_limits(self, high=None, low=None):
+        if low is None:
+            low = 4 * self._block_size
+
+        if high is None:
+            high = 8 * self._block_size
+
+        if low < 0:
+            raise ValueError("the limits must be positive")
+
+        if high < 0:
+            raise ValueError("the limits must be positive")
+
+        if low > high:
+            low = high
+
+        self._output_buffer_limit_low = low
+        self._output_buffer_limit_high = high
+
+    def get_write_buffer_limits(self):
+        return self._output_buffer_limit_low, self._output_buffer_limit_high
+
+    def get_write_buffer_size(self):
+        return len(self._write_buffer)
+
+    def set_protocol(self, proto):
+        self._protocol = proto
+        proto.connection_made(self)
+
+    def get_protocol(self):
+        return self._protocol
+
+    def pause_reading(self):
+        self._reading_paused = True
+
+    def resume_reading(self):
+        self._reading_paused = False
+        if self._protocol is not None:
+            self._protocol.data_received(b"".join(self._input_buffer))
```

Comment:
Don’t you need to clear the input buffer here, too?

Reply:
No, what I mean is: after calling ``data_received``, doesn’t it make sense to clear the ``_input_buffer``? Or is that clearing happening elsewhere?

Reply:
the input buffer is cleared in the next line <https://github.com/horazont/aioxmpp/pull/247/files/314cf831ac127d73dbc5330521ffea55d2455c5d#diff-14b03c7eccb3da4229e43ab3f717466dR126>. (The input buffer is a list of `bytes` objects).

Reply:
Uff… I’ll steal your statement about reading skills.